**Atributes to keep + description:**  
`trip_id`: identifier of the trip, not unique (there are as many trip_id as the length of the trip ie. number of stations in the trip, each stop in the trip has a number given by stop_sequence)    
`arrival_time`: scheduled (local) time of arrival at the stop (same as DEPARTURE_TIME if this is the start of the journey)  
`departure_time` : scheduled (local) time of departure at the stop  
`stop_id`: stop (station) identifier (FK), from stops.txt  
`stop_sequence`: sequence number of the stop on this trip id, starting at 1.  

Note about the attributes:  
If Departure_time == Arrival_time in the middle of a trip, this means that the train waits less that 1 minute (in theory) in the station before continuing. But all the stop_id in each trip is a station where the trains stops at.  
Stop_sequence: 1 = departure station, final_number=arrival station, all other numbers in the sequence=stations where the train stops at, but not arrival.  

Data cleaning notes:  
-ARRIVAL_TIME and DEPARTURE_TIME has some malformed date_times (more than 24h)  
-no nan beyond malformed arrival and departure times





In [1]:
import pandas as pd
#pd.options.display.max_rows = 10
#pd.options.mode.use_inf_as_na = True

In [85]:
filepath = "../data/stop_times.txt"
df = pd.read_csv(filepath)

C:\Users\Henrique\AppData\Local\Temp\ipykernel_8144\4111850311.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(filepath)


In [86]:
# print number of nan values in each column
print(df.isna().sum())

trip_id           0
arrival_time      0
departure_time    0
stop_id           0
stop_sequence     0
pickup_type       0
drop_off_type     0
dtype: int64


In [3]:
df.dtypes

trip_id           object
arrival_time      object
departure_time    object
stop_id           object
stop_sequence      int64
pickup_type        int64
drop_off_type      int64
dtype: object

In [7]:
df.head()

trip_id arrival_time departure_time  stop_id  stop_sequence  \
0  1.TA.37-1-Y-j20-1.1.H     10:00:00       10:00:00  8505380              1   
1  1.TA.37-1-Y-j20-1.1.H     13:00:00       13:00:00  8578641              2   
2  1.TA.37-2-Y-j20-1.1.H     15:12:00       15:12:00  8574796              1   
3  1.TA.37-2-Y-j20-1.1.H     16:04:00       16:04:00  8574935              2   
4  1.TA.37-3-Y-j20-1.1.H     10:28:00       10:28:00  8574935              1   

   pickup_type  drop_off_type  
0            0              0  
1            0              0  
2            0              0  
3            0              0  
4            0              0

In [10]:
# convert arrival_time and departure_time to datetime
df['arrival_time'] = pd.to_datetime(df['arrival_time'], format='%H:%M:%S', errors='coerce')
df['departure_time'] = pd.to_datetime(df['departure_time'], format='%H:%M:%S', errors='coerce')

In [11]:
# compute number of naT values in percent of total and drop naT values
malformed_percent = df.isna().sum() / len(df) * 100
df = df.dropna()
malformed_percent

trip_id           0.000000
arrival_time      3.234473
departure_time    3.255544
stop_id           0.000000
stop_sequence     0.000000
pickup_type       0.000000
drop_off_type     0.000000
dtype: float64

In [12]:
print(f"Number of rows: {len(df)}")

Number of rows: 11510548


In [17]:
df.describe()

stop_sequence   pickup_type  drop_off_type
count   1.151055e+07  1.151055e+07   1.151055e+07
mean    9.920480e+00  2.947644e-03   7.077856e-04
std     7.980645e+00  5.421214e-02   2.659482e-02
min     1.000000e+00  0.000000e+00   0.000000e+00
25%     3.000000e+00  0.000000e+00   0.000000e+00
50%     8.000000e+00  0.000000e+00   0.000000e+00
75%     1.400000e+01  0.000000e+00   0.000000e+00
max     6.600000e+01  1.000000e+00   1.000000e+00

Each row corresponds to a stop, there are 11,897,889 in total ie 11 mio approx.  
We need to group by trip_id to get number of trips.  
However, we only need to keep trips 15km away max from Zürich

In [19]:
df["trip_id"].nunique()

1112257

In [11]:
df.groupby('trip_id').size().sort_values(ascending=False)

trip_id
3.TA.16-623-A-j20-1.3.H       69
249.TA.27-11-j20-1.27.R       66
253.TA.27-11-j20-1.28.R       66
243.TA.27-11-j20-1.27.R       66
87.TA.27-11-j20-1.12.H        66
                              ..
3178.TA.61-101-Y-j20-1.1.H     2
3177.TA.88-146-Y-j20-1.1.H     2
3177.TA.61-333-Y-j20-1.1.H     2
3177.TA.61-281-Y-j20-1.1.H     2
999.TA.90-161-Y-j20-1.1.H      2
Length: 1132287, dtype: int64

Longest trip has 69 stops, shortest has 2 stops (start stop and arrival stop)

example below:

In [12]:
df.head(2)

trip_id arrival_time departure_time  stop_id  stop_sequence  \
0  1.TA.37-1-Y-j20-1.1.H     10:00:00       10:00:00  8505380              1   
1  1.TA.37-1-Y-j20-1.1.H     13:00:00       13:00:00  8578641              2   

   pickup_type  drop_off_type  
0            0              0  
1            0              0

This trip has two stops: departure (stop_id=8505380) and arrival (stop_id=8578641).  
Trip starts at 10h00 and trains arrive at 13h00  

In [20]:
df[df['trip_id'] == "1.TA.91-1-Y-j20-1.1.H"]

trip_id        arrival_time      departure_time  \
8   1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:17:00 1900-01-01 08:17:00   
9   1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:19:00 1900-01-01 08:19:00   
10  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:23:00 1900-01-01 08:23:00   
11  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:25:00 1900-01-01 08:25:00   
12  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:28:00 1900-01-01 08:29:00   
13  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:30:00 1900-01-01 08:30:00   
14  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:32:00 1900-01-01 08:32:00   
15  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:34:00 1900-01-01 08:34:00   
16  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:39:00 1900-01-01 08:40:00   
17  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:43:00 1900-01-01 08:43:00   
18  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:45:00 1900-01-01 08:45:00   
19  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:47:00 1900-01-01 08:47:00   
20  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:51:00 1900-01-01 08:51:00   
21  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:56:00 1900-01-01 08:56:00   
22  1.TA.91-1-Y-j20-1.1.H 1900-01-01 08:59:00 1900-01-01 08:59:00   
23  1.TA.91-1-Y-j20-1.1.H 1900-01-01 09:06:00 1900-01-01 09:06:00   

        stop_id  stop_sequence  pickup_type  drop_off_type  
8   8501300:0:8              1            0              0  
9       8501360              2            0              0  
10      8501361              3            0              0  
11      8501362              4            0              0  
12      8501370              5            0              0  
13      8501230              6            0              0  
14      8501308              7            0              0  
15      8501363              8            0              0  
16      8501371              9            0              0  
17      8501364             10            0              0  
18      8501357             11            0              0  
19      8501365             12            0              0  
20      8501366             13            0              0  
21      8501367             14            0              0  
22      8501368             15            0              0  
23      8501369             16            0              0

In [24]:
df[df['trip_id'] == "7.TA.91-1-Y-j20-1.1.H"]

trip_id        arrival_time      departure_time  \
24  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:17:00 1900-01-01 09:17:00   
25  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:19:00 1900-01-01 09:19:00   
26  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:23:00 1900-01-01 09:23:00   
27  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:25:00 1900-01-01 09:25:00   
28  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:28:00 1900-01-01 09:29:00   
29  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:30:00 1900-01-01 09:30:00   
30  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:32:00 1900-01-01 09:32:00   
31  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:34:00 1900-01-01 09:34:00   
32  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:39:00 1900-01-01 09:40:00   
33  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:43:00 1900-01-01 09:43:00   
34  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:45:00 1900-01-01 09:45:00   
35  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:47:00 1900-01-01 09:47:00   
36  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:51:00 1900-01-01 09:51:00   
37  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:56:00 1900-01-01 09:56:00   
38  7.TA.91-1-Y-j20-1.1.H 1900-01-01 09:59:00 1900-01-01 09:59:00   
39  7.TA.91-1-Y-j20-1.1.H 1900-01-01 10:06:00 1900-01-01 10:06:00   

        stop_id  stop_sequence  pickup_type  drop_off_type  
24  8501300:0:8              1            0              0  
25      8501360              2            0              0  
26      8501361              3            0              0  
27      8501362              4            0              0  
28      8501370              5            0              0  
29      8501230              6            0              0  
30      8501308              7            0              0  
31      8501363              8            0              0  
32      8501371              9            0              0  
33      8501364             10            0              0  
34      8501357             11            0              0  
35      8501365             12            0              0  
36      8501366             13            0              0  
37      8501367             14            0              0  
38      8501368             15            0              0  
39      8501369             16            0              0

What does this one mean ? Sometimes arrival_time = departure_time on the middle of the trip


# Analyze a known trip

In [25]:
df_stops = pd.read_csv("../data/stops.txt")
df_stops.head()

stop_id                      stop_name   stop_lat  stop_lon  location_type  \
0  1100006      Zell (Wiesental), Bahnhof  47.704632  7.847772            NaN   
1  1100008  Zell (Wiesental), Wilder Mann  47.710084  7.859648            NaN   
2  1100009     Zell (Wiesental), Grönland  47.713191  7.862909            NaN   
3  1100010                      Atzenbach  47.714624  7.872341            NaN   
4  1100011                Mambach, Brücke  47.728209  7.877470            NaN   

  parent_station  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

In [44]:
# all stops with "Collombey" in the name
df_stops[df_stops['stop_name'].str.contains("Collombey")]

stop_id                       stop_name   stop_lat  stop_lon  \
6006       8501221      Muraz (Collombey), collège  46.278389  6.929415   
6214   8501422:0:1                       Collombey  46.271254  6.948756   
6240       8501462                 Collombey-Muraz  46.270168  6.946016   
6696       8501983           Saillon, Collombeyres  46.171886  7.187268   
9442       8504459   Muraz (Collombey), Pré Géroux  46.288080  6.933215   
9460       8504473                Collombey, Barme  46.274874  6.941381   
9463       8504476           Collombey, Charmettes  46.272186  6.944651   
9464       8504478            Collombey, St Didier  46.268746  6.951846   
9511       8504497        Collombey, ch. du Verger  46.265802  6.946178   
9515       8504502    Collombey-le-Grand, Chapelle  46.281425  6.945944   
19795      8575520  Collombey, Les Neyres Chapelle  46.253070  6.934464   
23171      8579373     Muraz (Collombey), Millière  46.280916  6.924717   
23172      8579375               Collombey, église  46.267696  6.946223   
23176      8579382              Collombey-le-Grand  46.279624  6.947921   
23177      8579383             Collombey, Guinchet  46.274694  6.950202   
23178      8579384              Collombey, Corbier  46.272341  6.950939   
23179      8579386          Collombey, Les Vergers  46.270547  6.947454   
23180      8579387             Collombey, Le Foyer  46.275160  6.940914   
25300      8582424        Collombey, rte de la Fin  46.277830  6.949214   
25301      8582425        Collombey, Les Perraires  46.276669  6.941058   
25306      8582434    Collombey, Les Neyres Raccot  46.252237  6.938228   
25311      8582439           Collombey, Pré Géroux  46.290774  6.936746   
26006      8583387   Muraz (Collombey), Raffinerie  46.281356  6.932730   
26007      8583389    Collombey, Centre commercial  46.265107  6.954766   
26125      8583543    Muraz (Collombey), Pré Court  46.278035  6.936270   
26239      8583666        Collombey, Reutet-Bovéry  46.284548  6.942279   
26308      8583750  Muraz (Collombey), ch. Andains  46.284150  6.929981   

       location_type parent_station  
6006             NaN            NaN  
6214             NaN            NaN  
6240             NaN            NaN  
6696             NaN            NaN  
9442             NaN            NaN  
9460             NaN            NaN  
9463             NaN            NaN  
9464             NaN            NaN  
9511             NaN            NaN  
9515             NaN            NaN  
19795            NaN            NaN  
23171            NaN            NaN  
23172            NaN            NaN  
23176            NaN            NaN  
23177            NaN            NaN  
23178            NaN            NaN  
23179            NaN            NaN  
23180            NaN            NaN  
25300            NaN            NaN  
25301            NaN            NaN  
25306            NaN            NaN  
25311            NaN            NaN  
26006            NaN            NaN  
26007            NaN            NaN  
26125            NaN            NaN  
26239            NaN            NaN  
26308            NaN            NaN

In [28]:
df_stops[df_stops["stop_name"] == "St-Maurice"]

stop_id   stop_name   stop_lat  stop_lon  location_type  \
6196        8501403  St-Maurice  46.216372  7.002044            NaN   
6197    8501403:0:1  St-Maurice  46.216683  7.002044            NaN   
6198    8501403:0:2  St-Maurice  46.216620  7.002044            NaN   
6199    8501403:0:3  St-Maurice  46.216745  7.002044            NaN   
6200  8501403:0:3AB  St-Maurice  46.216807  7.002044            NaN   
6201    8501403:0:4  St-Maurice  46.216434  7.002044            NaN   
6202    8501403:0:6  St-Maurice  46.216496  7.002044            NaN   
6203       8501403P  St-Maurice  46.216372  7.002044            1.0   

     parent_station  
6196       8501403P  
6197       8501403P  
6198       8501403P  
6199       8501403P  
6200       8501403P  
6201       8501403P  
6202       8501403P  
6203            NaN

In [68]:
df_stops[df_stops["stop_name"] == "Martigny"]

stop_id stop_name   stop_lat  stop_lon  location_type  \
6275        8501500  Martigny  46.105825   7.07912            NaN   
6276    8501500:0:1  Martigny  46.105950   7.07912            NaN   
6277  8501500:0:1AB  Martigny  46.106386   7.07912            NaN   
6278  8501500:0:1CD  Martigny  46.106323   7.07912            NaN   
6279    8501500:0:2  Martigny  46.106012   7.07912            NaN   
6280    8501500:0:3  Martigny  46.105887   7.07912            NaN   
6281   8501500:0:40  Martigny  46.106136   7.07912            NaN   
6282   8501500:0:50  Martigny  46.106074   7.07912            NaN   
6283       8501500P  Martigny  46.105825   7.07912            1.0   

     parent_station  
6275       8501500P  
6276       8501500P  
6277       8501500P  
6278       8501500P  
6279       8501500P  
6280       8501500P  
6281       8501500P  
6282       8501500P  
6283            NaN

In [77]:
df_stops[df_stops["stop_name"] == "Monthey"]

stop_id stop_name   stop_lat  stop_lon  location_type parent_station
6213  8501421   Monthey  46.255654  6.955224            NaN            NaN

In [35]:
df[df['stop_id'] == "8501403"].head(20)

trip_id        arrival_time      departure_time  \
9769179   1.TA.30-100-Y-j20-1.1.H 1900-01-01 05:08:00 1900-01-01 05:08:00   
9769191   2.TA.30-100-Y-j20-1.1.H 1900-01-01 05:26:00 1900-01-01 05:26:00   
9769203   3.TA.30-100-Y-j20-1.2.H 1900-01-01 05:40:00 1900-01-01 05:40:00   
9769207   4.TA.30-100-Y-j20-1.2.H 1900-01-01 06:40:00 1900-01-01 06:40:00   
9769211   5.TA.30-100-Y-j20-1.2.H 1900-01-01 07:40:00 1900-01-01 07:40:00   
9769215   6.TA.30-100-Y-j20-1.2.H 1900-01-01 08:40:00 1900-01-01 08:40:00   
9769219   7.TA.30-100-Y-j20-1.2.H 1900-01-01 09:40:00 1900-01-01 09:40:00   
9769223   8.TA.30-100-Y-j20-1.2.H 1900-01-01 10:40:00 1900-01-01 10:40:00   
9769227   9.TA.30-100-Y-j20-1.2.H 1900-01-01 11:40:00 1900-01-01 11:40:00   
9769231  10.TA.30-100-Y-j20-1.2.H 1900-01-01 12:40:00 1900-01-01 12:40:00   
9769235  11.TA.30-100-Y-j20-1.2.H 1900-01-01 13:40:00 1900-01-01 13:40:00   
9769239  12.TA.30-100-Y-j20-1.2.H 1900-01-01 14:40:00 1900-01-01 14:40:00   
9769243  13.TA.30-100-Y-j20-1.2.H 1900-01-01 15:40:00 1900-01-01 15:40:00   
9769247  14.TA.30-100-Y-j20-1.2.H 1900-01-01 16:40:00 1900-01-01 16:40:00   
9769251  15.TA.30-100-Y-j20-1.2.H 1900-01-01 17:40:00 1900-01-01 17:40:00   
9769255  16.TA.30-100-Y-j20-1.2.H 1900-01-01 18:40:00 1900-01-01 18:40:00   
9769259  17.TA.30-100-Y-j20-1.2.H 1900-01-01 19:40:00 1900-01-01 19:40:00   
9769263  18.TA.30-100-Y-j20-1.2.H 1900-01-01 20:40:00 1900-01-01 20:40:00   
9769267  19.TA.30-100-Y-j20-1.2.H 1900-01-01 21:40:00 1900-01-01 21:40:00   
9769271  20.TA.30-100-Y-j20-1.2.H 1900-01-01 22:39:00 1900-01-01 22:39:00   

         stop_id  stop_sequence  pickup_type  drop_off_type  
9769179  8501403              1            0              0  
9769191  8501403              1            0              0  
9769203  8501403              1            0              0  
9769207  8501403              1            0              0  
9769211  8501403              1            0              0  
9769215  8501403              1            0              0  
9769219  8501403              1            0              0  
9769223  8501403              1            0              0  
9769227  8501403              1            0              0  
9769231  8501403              1            0              0  
9769235  8501403              1            0              0  
9769239  8501403              1            0              0  
9769243  8501403              1            0              0  
9769247  8501403              1            0              0  
9769251  8501403              1            0              0  
9769255  8501403              1            0              0  
9769259  8501403              1            0              0  
9769263  8501403              1            0              0  
9769267  8501403              1            0              0  
9769271  8501403              1            0              0

## St-Maurice

In [38]:
ex_trip = df[df["trip_id"] == "1.TA.30-100-Y-j20-1.1.H"]

In [41]:
# print all the stop names of the trip
for i, stop_id in enumerate(ex_trip["stop_id"]):
    print("Stop name {}: {}".format(i, df_stops[df_stops["stop_id"] == stop_id]["stop_name"].values[0]))

Stop name 0: St-Maurice
Stop name 1: Bex
Stop name 2: Aigle
Stop name 3: Roche VD
Stop name 4: Villeneuve
Stop name 5: Veytaux-Chillon
Stop name 6: Territet
Stop name 7: Montreux
Stop name 8: Clarens
Stop name 9: Burier
Stop name 10: La Tour-de-Peilz
Stop name 11: Vevey


**Hypothesis**  
When arrival_time==departure_time DURING a trip for a stop_id, then it means that the train doesn't stop at this stop_id, but passes through it.

## Collombey

In [65]:
df[df["stop_id"] == "8501422:0:1"].head()

trip_id        arrival_time      departure_time  \
10657893   1.TA.66-130-j20-1.1.H 1900-01-01 22:43:00 1900-01-01 22:43:00   
10657900   2.TA.66-130-j20-1.2.H 1900-01-01 23:43:00 1900-01-01 23:43:00   
10657909  79.TA.66-130-j20-1.5.H 1900-01-01 09:10:00 1900-01-01 09:10:00   
10657937  81.TA.66-130-j20-1.5.H 1900-01-01 09:10:00 1900-01-01 09:10:00   
10657965  84.TA.66-130-j20-1.5.H 1900-01-01 10:10:00 1900-01-01 10:10:00   

              stop_id  stop_sequence  pickup_type  drop_off_type  
10657893  8501422:0:1              6            0              0  
10657900  8501422:0:1              6            0              0  
10657909  8501422:0:1              6            0              0  
10657937  8501422:0:1              6            0              0  
10657965  8501422:0:1              6            0              0

In [60]:
ex_trip2 = df[df["trip_id"] == "1.TA.66-130-j20-1.1.H"]
# print all the stop names of the trip
for i, stop_id in enumerate(ex_trip2["stop_id"]):
    print("Stop name {}: {}".format(i, df_stops[df_stops["stop_id"] == stop_id]["stop_name"].values[0]))

Stop name 0: St-Gingolph (Suisse)
Stop name 1: Bouveret
Stop name 2: Les Evouettes
Stop name 3: Vouvry
Stop name 4: Vionnaz
Stop name 5: Collombey
Stop name 6: Monthey


In [61]:
ex_trip2

trip_id        arrival_time      departure_time  \
10657888  1.TA.66-130-j20-1.1.H 1900-01-01 22:27:00 1900-01-01 22:27:00   
10657889  1.TA.66-130-j20-1.1.H 1900-01-01 22:32:00 1900-01-01 22:32:00   
10657890  1.TA.66-130-j20-1.1.H 1900-01-01 22:34:00 1900-01-01 22:34:00   
10657891  1.TA.66-130-j20-1.1.H 1900-01-01 22:37:00 1900-01-01 22:37:00   
10657892  1.TA.66-130-j20-1.1.H 1900-01-01 22:40:00 1900-01-01 22:40:00   
10657893  1.TA.66-130-j20-1.1.H 1900-01-01 22:43:00 1900-01-01 22:43:00   
10657894  1.TA.66-130-j20-1.1.H 1900-01-01 22:49:00 1900-01-01 22:49:00   

              stop_id  stop_sequence  pickup_type  drop_off_type  
10657888  8501428:0:1              1            0              0  
10657889  8501426:0:1              2            0              0  
10657890  8501425:0:1              3            0              0  
10657891  8501424:0:1              4            0              0  
10657892  8501423:0:1              5            0              0  
10657893  8501422:0:1              6            0              0  
10657894      8501421              7            0              0

In [66]:
ex_trip3 = df[df["trip_id"] == "79.TA.66-130-j20-1.5.H"]
# print all the stop names of the trip
for i, stop_id in enumerate(ex_trip2["stop_id"]):
    print("Stop name {}: {}".format(i, df_stops[df_stops["stop_id"] == stop_id]["stop_name"].values[0]))

Stop name 0: St-Gingolph (Suisse)
Stop name 1: Bouveret
Stop name 2: Les Evouettes
Stop name 3: Vouvry
Stop name 4: Vionnaz
Stop name 5: Collombey
Stop name 6: Monthey


In [67]:
ex_trip3

trip_id        arrival_time      departure_time  \
10657904  79.TA.66-130-j20-1.5.H 1900-01-01 08:52:00 1900-01-01 08:52:00   
10657905  79.TA.66-130-j20-1.5.H 1900-01-01 08:57:00 1900-01-01 08:59:00   
10657906  79.TA.66-130-j20-1.5.H 1900-01-01 09:01:00 1900-01-01 09:01:00   
10657907  79.TA.66-130-j20-1.5.H 1900-01-01 09:04:00 1900-01-01 09:04:00   
10657908  79.TA.66-130-j20-1.5.H 1900-01-01 09:07:00 1900-01-01 09:07:00   
10657909  79.TA.66-130-j20-1.5.H 1900-01-01 09:10:00 1900-01-01 09:10:00   
10657910  79.TA.66-130-j20-1.5.H 1900-01-01 09:16:00 1900-01-01 09:17:00   
10657911  79.TA.66-130-j20-1.5.H 1900-01-01 09:19:00 1900-01-01 09:19:00   
10657912  79.TA.66-130-j20-1.5.H 1900-01-01 09:24:00 1900-01-01 09:36:00   
10657913  79.TA.66-130-j20-1.5.H 1900-01-01 09:40:00 1900-01-01 09:40:00   
10657914  79.TA.66-130-j20-1.5.H 1900-01-01 09:43:00 1900-01-01 09:43:00   
10657915  79.TA.66-130-j20-1.5.H 1900-01-01 09:49:00 1900-01-01 09:50:00   
10657916  79.TA.66-130-j20-1.5.H 1900-01-01 09:53:00 1900-01-01 09:53:00   
10657917  79.TA.66-130-j20-1.5.H 1900-01-01 09:56:00 1900-01-01 09:57:00   
10657918  79.TA.66-130-j20-1.5.H 1900-01-01 10:00:00 1900-01-01 10:00:00   
10657919  79.TA.66-130-j20-1.5.H 1900-01-01 10:02:00 1900-01-01 10:02:00   
10657920  79.TA.66-130-j20-1.5.H 1900-01-01 10:05:00 1900-01-01 10:05:00   
10657921  79.TA.66-130-j20-1.5.H 1900-01-01 10:07:00 1900-01-01 10:07:00   
10657922  79.TA.66-130-j20-1.5.H 1900-01-01 10:13:00 1900-01-01 10:14:00   
10657923  79.TA.66-130-j20-1.5.H 1900-01-01 10:17:00 1900-01-01 10:17:00   
10657924  79.TA.66-130-j20-1.5.H 1900-01-01 10:24:00 1900-01-01 10:25:00   
10657925  79.TA.66-130-j20-1.5.H 1900-01-01 10:28:00 1900-01-01 10:28:00   
10657926  79.TA.66-130-j20-1.5.H 1900-01-01 10:32:00 1900-01-01 10:33:00   
10657927  79.TA.66-130-j20-1.5.H 1900-01-01 10:36:00 1900-01-01 10:36:00   
10657928  79.TA.66-130-j20-1.5.H 1900-01-01 10:39:00 1900-01-01 10:39:00   
10657929  79.TA.66-130-j20-1.5.H 1900-01-01 10:42:00 1900-01-01 10:42:00   
10657930  79.TA.66-130-j20-1.5.H 1900-01-01 10:48:00 1900-01-01 10:49:00   
10657931  79.TA.66-130-j20-1.5.H 1900-01-01 10:56:00 1900-01-01 10:56:00   

              stop_id  stop_sequence  pickup_type  drop_off_type  
10657904  8501428:0:1              1            0              0  
10657905  8501426:0:1              2            0              0  
10657906  8501425:0:1              3            0              0  
10657907  8501424:0:1              4            0              0  
10657908  8501423:0:1              5            0              0  
10657909  8501422:0:1              6            0              0  
10657910      8501421              7            0              0  
10657911  8501420:0:1              8            0              0  
10657912  8501403:0:3              9            0              0  
10657913  8501404:0:2             10            0              0  
10657914  8501405:0:2             11            0              0  
10657915  8501500:0:2             12            0              0  
10657916  8501501:0:2             13            0              0  
10657917  8501502:0:2             14            0              0  
10657918  8501503:0:2             15            0              0  
10657919  8501504:0:1             16            0              0  
10657920  8501505:0:2             17            0              0  
10657921  8501510:0:1             18            0              0  
10657922  8501506:0:2             19            0              0  
10657923  8501507:0:1             20            0              0  
10657924  8501509:0:2             21            0              0  
10657925  8501600:0:1             22            0              0  
10657926  8501601:0:2             23            0              0  
10657927  8501602:0:2             24            0              0  
10657928  8501603:0:4             25            0              0  
10657929  8501604:0:3             26            0              0  
10657930  8

## Martigny

In [71]:
df[df["stop_id"] == "8501500"].head(20)

trip_id        arrival_time      departure_time  \
9818682   1.TA.30-515-Y-j20-1.1.H 1900-01-01 05:45:00 1900-01-01 05:45:00   
9818686  57.TA.30-515-Y-j20-1.1.H 1900-01-01 06:09:00 1900-01-01 06:09:00   
9818690  58.TA.30-515-Y-j20-1.1.H 1900-01-01 06:59:00 1900-01-01 06:59:00   
9818694  59.TA.30-515-Y-j20-1.1.H 1900-01-01 08:09:00 1900-01-01 08:09:00   
9818698  60.TA.30-515-Y-j20-1.1.H 1900-01-01 09:09:00 1900-01-01 09:09:00   
9818702  61.TA.30-515-Y-j20-1.1.H 1900-01-01 10:09:00 1900-01-01 10:09:00   
9818706  62.TA.30-515-Y-j20-1.1.H 1900-01-01 11:09:00 1900-01-01 11:09:00   
9818710  63.TA.30-515-Y-j20-1.1.H 1900-01-01 12:09:00 1900-01-01 12:09:00   
9818714  64.TA.30-515-Y-j20-1.1.H 1900-01-01 13:09:00 1900-01-01 13:09:00   
9818718  65.TA.30-515-Y-j20-1.1.H 1900-01-01 14:09:00 1900-01-01 14:09:00   
9818722  66.TA.30-515-Y-j20-1.1.H 1900-01-01 15:09:00 1900-01-01 15:09:00   
9818726  67.TA.30-515-Y-j20-1.1.H 1900-01-01 16:09:00 1900-01-01 16:09:00   
9818730  68.TA.30-515-Y-j20-1.1.H 1900-01-01 17:09:00 1900-01-01 17:09:00   
9818734  69.TA.30-515-Y-j20-1.1.H 1900-01-01 18:09:00 1900-01-01 18:09:00   
9818738  70.TA.30-515-Y-j20-1.1.H 1900-01-01 19:09:00 1900-01-01 19:09:00   
9818742  71.TA.30-515-Y-j20-1.1.H 1900-01-01 20:09:00 1900-01-01 20:09:00   
9818746  72.TA.30-515-Y-j20-1.1.H 1900-01-01 21:09:00 1900-01-01 21:09:00   
9818750  73.TA.30-515-Y-j20-1.1.H 1900-01-01 22:09:00 1900-01-01 22:09:00   
9818754  74.TA.30-515-Y-j20-1.1.H 1900-01-01 23:09:00 1900-01-01 23:09:00   
9818760   2.TA.30-515-Y-j20-1.2.H 1900-01-01 07:27:00 1900-01-01 07:27:00   

         stop_id  stop_sequence  pickup_type  drop_off_type  
9818682  8501500              4            0              0  
9818686  8501500              4            0              0  
9818690  8501500              4            0              0  
9818694  8501500              4            0              0  
9818698  8501500              4            0              0  
9818702  8501500              4            0              0  
9818706  8501500              4            0              0  
9818710  8501500              4            0              0  
9818714  8501500              4            0              0  
9818718  8501500              4            0              0  
9818722  8501500              4            0              0  
9818726  8501500              4            0              0  
9818730  8501500              4            0              0  
9818734  8501500              4            0              0  
9818738  8501500              4            0              0  
9818742  8501500              4            0              0  
9818746  8501500              4            0              0  
9818750  8501500              4            0              0  
9818754  8501500              4            0              0  
9818760  8501500              2            0              0

**This is a direct:**

In [72]:
ex_martigny = df[df["trip_id"] == "2.TA.30-515-Y-j20-1.2.H"]
# print all the stop names of the trip
for i, stop_id in enumerate(ex_martigny["stop_id"]):
    print("Stop name {}: {}".format(i, df_stops[df_stops["stop_id"] == stop_id]["stop_name"].values[0]))

Stop name 0: St-Maurice
Stop name 1: Martigny


In [75]:
ex_martigny

trip_id        arrival_time      departure_time  \
9818759  2.TA.30-515-Y-j20-1.2.H 1900-01-01 07:07:00 1900-01-01 07:07:00   
9818760  2.TA.30-515-Y-j20-1.2.H 1900-01-01 07:27:00 1900-01-01 07:27:00   

         stop_id  stop_sequence  pickup_type  drop_off_type  
9818759  8501403              1            0              0  
9818760  8501500              2            0              0

**This is not a direct:**

In [74]:
ex_martigny2 = df[df["trip_id"] == "1.TA.30-515-Y-j20-1.1.H"]
# print all the stop names of the trip
for i, stop_id in enumerate(ex_martigny2["stop_id"]):
    print("Stop name {}: {}".format(i, df_stops[df_stops["stop_id"] == stop_id]["stop_name"].values[0]))

Stop name 0: St-Maurice
Stop name 1: Evionnaz
Stop name 2: Vernayaz
Stop name 3: Martigny


In [76]:
ex_martigny2

trip_id        arrival_time      departure_time  \
9818679  1.TA.30-515-Y-j20-1.1.H 1900-01-01 05:17:00 1900-01-01 05:17:00   
9818680  1.TA.30-515-Y-j20-1.1.H 1900-01-01 05:29:00 1900-01-01 05:29:00   
9818681  1.TA.30-515-Y-j20-1.1.H 1900-01-01 05:37:00 1900-01-01 05:37:00   
9818682  1.TA.30-515-Y-j20-1.1.H 1900-01-01 05:45:00 1900-01-01 05:45:00   

         stop_id  stop_sequence  pickup_type  drop_off_type  
9818679  8501403              1            0              0  
9818680  8501404              2            0              0  
9818681  8501405              3            0              0  
9818682  8501500              4            0              0

**Hypothesis:**  
If departure_time=arrival_time during a trip, it means that the train doesn't wait a lot (less than 1 min in theory) in the stop before starting over. The above example shows that the hypothesis above is wrong!.

## Monthey

In [78]:
df[df["stop_id"] == "8501421"].head(20)

trip_id        arrival_time      departure_time  \
5484405    22.TA.24-103-j20-1.3.R 1900-01-01 01:03:00 1900-01-01 01:03:00   
5484432    23.TA.24-103-j20-1.3.R 1900-01-01 02:03:00 1900-01-01 02:03:00   
5484459    24.TA.24-103-j20-1.3.R 1900-01-01 03:03:00 1900-01-01 03:03:00   
5484486    25.TA.24-103-j20-1.3.R 1900-01-01 04:03:00 1900-01-01 04:03:00   
10646433   1.TA.63-83-Y-j20-1.1.H 1900-01-01 14:35:00 1900-01-01 14:35:00   
10657894    1.TA.66-130-j20-1.1.H 1900-01-01 22:49:00 1900-01-01 22:49:00   
10657901    2.TA.66-130-j20-1.2.H 1900-01-01 23:48:00 1900-01-01 23:49:00   
10657910   79.TA.66-130-j20-1.5.H 1900-01-01 09:16:00 1900-01-01 09:17:00   
10657938   81.TA.66-130-j20-1.5.H 1900-01-01 09:16:00 1900-01-01 09:17:00   
10657966   84.TA.66-130-j20-1.5.H 1900-01-01 10:16:00 1900-01-01 10:17:00   
10657994   86.TA.66-130-j20-1.5.H 1900-01-01 10:16:00 1900-01-01 10:17:00   
10658022   90.TA.66-130-j20-1.5.H 1900-01-01 20:16:00 1900-01-01 20:17:00   
10658050  101.TA.66-130-j20-1.5.H 1900-01-01 21:16:00 1900-01-01 21:17:00   
10658078  111.TA.66-130-j20-1.5.H 1900-01-01 21:16:00 1900-01-01 21:17:00   
10658106  144.TA.66-130-j20-1.5.H 1900-01-01 22:16:00 1900-01-01 22:17:00   
10658134  190.TA.66-130-j20-1.5.H 1900-01-01 11:16:00 1900-01-01 11:17:00   
10658162  193.TA.66-130-j20-1.5.H 1900-01-01 12:16:00 1900-01-01 12:17:00   
10658190  195.TA.66-130-j20-1.5.H 1900-01-01 12:16:00 1900-01-01 12:17:00   
10658218  218.TA.66-130-j20-1.5.H 1900-01-01 16:16:00 1900-01-01 16:17:00   
10658252  272.TA.66-130-j20-1.8.H 1900-01-01 23:17:00 1900-01-01 23:17:00   

          stop_id  stop_sequence  pickup_type  drop_off_type  
5484405   8501421              3            1              0  
5484432   8501421              3            1              0  
5484459   8501421              3            1              0  
5484486   8501421              3            1              0  
10646433  8501421              2            0              0  
10657894  8501421              7            0              0  
10657901  8501421              7            0              0  
10657910  8501421              7            0              0  
10657938  8501421              7            0              0  
10657966  8501421              7            0              0  
10657994  8501421              7            0              0  
10658022  8501421              7            0              0  
10658050  8501421              7            0              0  
10658078  8501421              7            0              0  
10658106  8501421              7            0              0  
10658134  8501421              7            0              0  
10658162  8501421              7            0              0  
10658190  8501421              7            0              0  
10658218  8501421              7            0              0  
10658252  8501421              1            0              0

In [81]:
ex_monthey1 = df[df["trip_id"] == "79.TA.66-130-j20-1.5.H"]
# print all the stop names of the trip
for i, stop_id in enumerate(ex_monthey1["stop_id"]):
    print("Stop name {}: {}".format(i, df_stops[df_stops["stop_id"] == stop_id]["stop_name"].values[0]))

Stop name 0: St-Gingolph (Suisse)
Stop name 1: Bouveret
Stop name 2: Les Evouettes
Stop name 3: Vouvry
Stop name 4: Vionnaz
Stop name 5: Collombey
Stop name 6: Monthey
Stop name 7: Massongex
Stop name 8: St-Maurice
Stop name 9: Evionnaz
Stop name 10: Vernayaz
Stop name 11: Martigny
Stop name 12: Charrat-Fully
Stop name 13: Saxon
Stop name 14: Riddes
Stop name 15: Chamoson
Stop name 16: Ardon
Stop name 17: Châteauneuf-Conthey
Stop name 18: Sion
Stop name 19: St-Léonard
Stop name 20: Sierre/Siders
Stop name 21: Salgesch
Stop name 22: Leuk
Stop name 23: Turtmann
Stop name 24: Gampel-Steg
Stop name 25: Raron
Stop name 26: Visp
Stop name 27: Brig


In [82]:
ex_monthey1

trip_id        arrival_time      departure_time  \
10657904  79.TA.66-130-j20-1.5.H 1900-01-01 08:52:00 1900-01-01 08:52:00   
10657905  79.TA.66-130-j20-1.5.H 1900-01-01 08:57:00 1900-01-01 08:59:00   
10657906  79.TA.66-130-j20-1.5.H 1900-01-01 09:01:00 1900-01-01 09:01:00   
10657907  79.TA.66-130-j20-1.5.H 1900-01-01 09:04:00 1900-01-01 09:04:00   
10657908  79.TA.66-130-j20-1.5.H 1900-01-01 09:07:00 1900-01-01 09:07:00   
10657909  79.TA.66-130-j20-1.5.H 1900-01-01 09:10:00 1900-01-01 09:10:00   
10657910  79.TA.66-130-j20-1.5.H 1900-01-01 09:16:00 1900-01-01 09:17:00   
10657911  79.TA.66-130-j20-1.5.H 1900-01-01 09:19:00 1900-01-01 09:19:00   
10657912  79.TA.66-130-j20-1.5.H 1900-01-01 09:24:00 1900-01-01 09:36:00   
10657913  79.TA.66-130-j20-1.5.H 1900-01-01 09:40:00 1900-01-01 09:40:00   
10657914  79.TA.66-130-j20-1.5.H 1900-01-01 09:43:00 1900-01-01 09:43:00   
10657915  79.TA.66-130-j20-1.5.H 1900-01-01 09:49:00 1900-01-01 09:50:00   
10657916  79.TA.66-130-j20-1.5.H 1900-01-01 09:53:00 1900-01-01 09:53:00   
10657917  79.TA.66-130-j20-1.5.H 1900-01-01 09:56:00 1900-01-01 09:57:00   
10657918  79.TA.66-130-j20-1.5.H 1900-01-01 10:00:00 1900-01-01 10:00:00   
10657919  79.TA.66-130-j20-1.5.H 1900-01-01 10:02:00 1900-01-01 10:02:00   
10657920  79.TA.66-130-j20-1.5.H 1900-01-01 10:05:00 1900-01-01 10:05:00   
10657921  79.TA.66-130-j20-1.5.H 1900-01-01 10:07:00 1900-01-01 10:07:00   
10657922  79.TA.66-130-j20-1.5.H 1900-01-01 10:13:00 1900-01-01 10:14:00   
10657923  79.TA.66-130-j20-1.5.H 1900-01-01 10:17:00 1900-01-01 10:17:00   
10657924  79.TA.66-130-j20-1.5.H 1900-01-01 10:24:00 1900-01-01 10:25:00   
10657925  79.TA.66-130-j20-1.5.H 1900-01-01 10:28:00 1900-01-01 10:28:00   
10657926  79.TA.66-130-j20-1.5.H 1900-01-01 10:32:00 1900-01-01 10:33:00   
10657927  79.TA.66-130-j20-1.5.H 1900-01-01 10:36:00 1900-01-01 10:36:00   
10657928  79.TA.66-130-j20-1.5.H 1900-01-01 10:39:00 1900-01-01 10:39:00   
10657929  79.TA.66-130-j20-1.5.H 1900-01-01 10:42:00 1900-01-01 10:42:00   
10657930  79.TA.66-130-j20-1.5.H 1900-01-01 10:48:00 1900-01-01 10:49:00   
10657931  79.TA.66-130-j20-1.5.H 1900-01-01 10:56:00 1900-01-01 10:56:00   

              stop_id  stop_sequence  pickup_type  drop_off_type  
10657904  8501428:0:1              1            0              0  
10657905  8501426:0:1              2            0              0  
10657906  8501425:0:1              3            0              0  
10657907  8501424:0:1              4            0              0  
10657908  8501423:0:1              5            0              0  
10657909  8501422:0:1              6            0              0  
10657910      8501421              7            0              0  
10657911  8501420:0:1              8            0              0  
10657912  8501403:0:3              9            0              0  
10657913  8501404:0:2             10            0              0  
10657914  8501405:0:2             11            0              0  
10657915  8501500:0:2             12            0              0  
10657916  8501501:0:2             13            0              0  
10657917  8501502:0:2             14            0              0  
10657918  8501503:0:2             15            0              0  
10657919  8501504:0:1             16            0              0  
10657920  8501505:0:2             17            0              0  
10657921  8501510:0:1             18            0              0  
10657922  8501506:0:2             19            0              0  
10657923  8501507:0:1             20            0              0  
10657924  8501509:0:2             21            0              0  
10657925  8501600:0:1             22            0              0  
10657926  8501601:0:2             23            0              0  
10657927  8501602:0:2             24            0              0  
10657928  8501603:0:4             25            0              0  
10657929  8501604:0:3             26            0              0  
10657930  8

In [84]:
df_stops[df_stops["stop_id"] == "8501403:0:3"]

stop_id   stop_name   stop_lat  stop_lon  location_type  \
6199  8501403:0:3  St-Maurice  46.216745  7.002044            NaN   

     parent_station  
6199       8501403P

As expected, there is a wait of 12 min in St-Maurice.  
arrival_time=09:24:00  
departure_time=09:36:00